File: src/main/java/net/felsstudio/fels/Modules/ftypes/ftypes.java
package main.java.net.felsstudio.fels.Modules.ftypes;

import main.java.net.felsstudio.fels.Modules.Module;
import main.java.net.felsstudio.fels.lib.MapValue;
import main.java.net.felsstudio.fels.lib.*;

public final class ftypes implements Module {

    @Override
    public void init() {
        final MapValue map = new MapValue(30);

        map.set("OBJECT", NumberValue.of(Types.OBJECT));
        map.set("NUMBER", NumberValue.of(Types.NUMBER));
        map.set("STRING", NumberValue.of(Types.STRING));
        map.set("ARRAY", NumberValue.of(Types.ARRAY));
        map.set("MAP", NumberValue.of(Types.MAP));
        map.set("FUNCTION", NumberValue.of(Types.FUNCTION));
        map.set("CLASS",NumberValue.of(Types.CLASS));

        map.set("typeToString",args -> new StringValue(Types.typeToString(args[0].asInt())));

        map.set("toByte", args -> NumberValue.of((byte)args[0].asInt()));
        map.set("toShort", args -> NumberValue.of((short)args[0].asInt()));
        map.set("toInt", args -> NumberValue.of(args[0].asInt()));
        map.set("toLong", args -> NumberValue.of((long)args[0].asNumber()));
        map.set("toFloat", args -> NumberValue.of((float)args[0].asNumber()));
        map.set("toDouble", args -> NumberValue.of(args[0].asNumber()));

        Variables.define("ftypes", map);
    }
}

File: src/net/felsstudio/fels/Modules/ftypes/ftypes.kt
package net.felsstudio.fels.Modules.ftypes

import net.felsstudio.fels.Modules.Module
import net.felsstudio.fels.lib.*

class ftypes : Module {
    override fun init() {
        Variables.set("OBJECT", NumberValue.of(Types.OBJECT))
        Variables.set("NUMBER", NumberValue.of(Types.NUMBER))
        Variables.set("STRING", NumberValue.of(Types.STRING))
        Variables.set("ARRAY", NumberValue.of(Types.ARRAY))
        Variables.set("MAP", NumberValue.of(Types.MAP))
        Variables.set("FUNCTION", NumberValue.of(Types.FUNCTION))

        Functions.set("typeOf") { args: Array<Value> ->
            NumberValue.of(
                args[0].type()
            )
        }
        Functions.set("toString") { args: Array<Value> ->
            StringValue(
                args[0].asString()
            )
        }
        Functions.set("toNumber") { args: Array<Value> ->
            NumberValue.of(
                args[0].asNumber()
            )
        }

        Functions.set("toByte") { args: Array<Value> ->
            NumberValue.of(
                args[0].asInt().toByte().toInt()
            )
        }
        Functions.set("toShort") { args: Array<Value> ->
            NumberValue.of(
                args[0].asInt().toShort().toInt()
            )
        }
        Functions.set("toInt") { args: Array<Value> ->
            NumberValue.of(
                args[0].asInt()
            )
        }
        Functions.set("toLong") { args: Array<Value> ->
            NumberValue.of(
                args[0].asNumber().toLong()
            )
        }
        Functions.set("toFloat") { args: Array<Value> ->
            NumberValue.of(
                args[0].asNumber().toFloat()
            )
        }
        Functions.set("toDouble") { args: Array<Value> ->
            NumberValue.of(
                args[0].asNumber()
            )
        }
    }
}